kimchi/src/circuits/argument.rs

```diff
@@ -0,0 +1,58 @@
+//! An argument is simply a number of constraints,
+//! which we want to enforce on all points of the domain.
+//! Both the permutation and the plookup arguments fit this type.
+//! Gates can be seen as filtered arguments,
+//! which apply only in some points (rows) of the domain.
+
+use crate::{alphas::Alphas, circuits::expr::prologue::*};
+use ark_ff::FftField;
+use serde::{Deserialize, Serialize};
+
+use super::gate::GateType;
+
+/// A constraint type represents a polynomial that will be part of the final equation f (the circuit equation)
+#[derive(PartialEq, Eq, Clone, Copy, Hash, Debug, Serialize, Deserialize)]
+pub enum ArgumentType {
+    /// Gates in the PLONK constraint system.
+    /// As gates are mutually exclusive (a single gate is set per row),
+    /// we can reuse the same powers of alpha across gates.
+    Gate(GateType),
+    /// The permutation argument
+    Permutation,
+    /// The lookup argument
+    Lookup,
+}
+
+/// The interface for a minimal argument implementation.
+pub trait Argument {
+    /// The field being used for this argument.
+    type Field: FftField;
```

Comment:
Is there a reason to do this vs. an `Argument<F: FftField>` trait bound? It currently feels redundant, given that we do an `impl<F: FftField> Argument for Foo` everywhere.

Reply:
That's what I did first, but I couldn't make it work for some reason. I admit that I don't remember what was the issue, but it'd be nice to make it work.